libtock/alarm_timer.c

```diff
@@ -106,6 +122,79 @@ int alarm_at(uint32_t reference, uint32_t dt, subscribe_upcall cb, void* ud, ala
   return RETURNCODE_SUCCESS;
 }
 
+// Holds original callback and user data. Also keeps bookkeeping for the
+// intermediate alarm overflows.
+typedef struct overflow_ud {
+  // How many times the alarm has left to overflow before reaching target time
+  int overflows_left;
+  // Number of remaining ticks after last full alarm overflow
+  int remainder_ticks;
+  // Original user data passed to `timer_in`
+  void* original_ud;
+  // Original callback passed to `timer_in`
+  subscribe_upcall* original_cb;
+  alarm_t* alarm;
+} overflow_ud_t;
+
+// The intermediate callback that handles overflows of alarm. This is used by
+// `timer_in` to handle cases where `alarm_at` would normally be unable to
+// support the whole timer length. `alarm_at` can only keep track of up to 2^32
+// ticks.
+static void overflow_callback(int                          last_timer_fire_time,
+                              __attribute__ ((unused)) int unused1,
+                              __attribute__ ((unused)) int unused2,
+                              void*                        overflow_ud) {
+  overflow_ud_t* ud = (overflow_ud_t*)overflow_ud;
+
+  if (ud->overflows_left == 0) {
+    // no overflows left, schedule last alarm with original callback
+    alarm_at(
+      last_timer_fire_time,
+      ud->remainder_ticks,
+      ud->original_cb,
+      ud->original_ud,
+      ud->alarm
+      );
+  } else {
+    // schedule next intermediate alarm that will overflow
+    ud->overflows_left--;
+
+    alarm_at(
+      last_timer_fire_time,
+      UINT_MAX,
```

Comment:
Is it alright to use `UINT_MAX` for the max value of `uint32_t`? They have the same value for me, but maybe not for everyone?

Reply:
Could you use `UINT32_MAX`?

Reply:
Yes, just fixed it. I tried that and it didn't compile earlier, not sure why.